### ISSUE TITLE
`@defer` does not work on fragment expansion, which is required on Relay 11

### ISSUE THREAD

**Issue by fazo96** [rmosolgo/graphql-ruby#3495]
**Describe the bug**

The defer keyword is ignored on fragment expansion.

**Versions**

`graphql` version: 15.5
`rails`: 6
`relay`: 11.0.2

**GraphQL schema**

I am using Relay 11 as client and the Defer extension from graphql-pro

According to people on the `#relay` channel in the GraphQL Slack server, Relay only supports `@defer` on fragment expansion. However in practice I found it works in other cases:

**GraphQL query**

These examples are not real, but they should be enough to understand the issue. I wrote them to condense the actual queries I have been running in my project

This is a query with `@defer` that works fine with graphql-pro, even though it's not correct usage of Relay apparently. This is not a graphql-ruby problem, I am just adding it for context.

```graphql
query WorkingFully1 {
  products {
    id
    title @defer(label: 'ProductTitle')
  }
}
```

The response is a proper multipart mixed response and Relay handles it perfectly

This query is also handled perfectly by the server, but unfortunately Relay does not handle the deferred response and discards the `products` array data. This is not a graphql-ruby problem and I am adding it just for context.

```graphql
query WorkingOnTheServer1 {
  products @defer(label: 'Products') {
    id
    title
  }
}
```

However, this next kind of query does not work at all. Relay sends it properly, but the `@defer` seems to be ignored by graphql-ruby and the data comes back as if it wasn't there. Relay ignores the data in the fragment, I guess because it expects it to be deferred. This is preventing me from taking advantage of `@defer` properly in Relay to defer the loading of some array-heavy objects I have.

```graphql
query Broken1 {
  warehouse(id: 1) {
    ...Products @defer(label: "Products")
  }
}

fragment Products on Warehouse {
  products { id title }
}
```

**Steps to reproduce**

Run a query similar to `Broken1` using Relay

**Expected behavior**

The server defers the data in the expanded fragment, passing `label` properly so that Relay can figure out how to merge it with the initial data (see https://github.com/rmosolgo/graphql-ruby/issues/3454), just like for the other two examples.

**Actual behavior**

`@defer` is ignored on the fragment expansion, making Relay discard its contents in the response.

**Additional context**

See examples queries above. Unfortunately Relay has zero docs on this so we have to go blind! The only thing I managed to gather on the Slack server is "defer only works on fragment expansion"

Also see https://github.com/rmosolgo/graphql-ruby/issues/3454#issuecomment-847263195

**Comment by rmosolgo** [comment]
Thanks for expanding this bug report a bit, @fazo96. I think it should be fine to make `graphql-pro` handle `@defer` on fragment spreads. I'll take a look and follow up here.

**Comment by rmosolgo** [comment]
It looks like this is a shortcoming in GraphQL-Ruby's execution code. GraphQL-Ruby resolves selections in a two-step approach: 

- First, merge all selected fields into a single set 
- Then, execute each field 

But that doesn't work when it comes to applying directives to fragment spreads, because that information is lost during the first step. (Directives on fragment spreads are ignored entirely.) Here's the problem code: 

https://github.com/rmosolgo/graphql-ruby/blob/9ecc20efadca8d499fe85e56748ab002f44b774b/lib/graphql/execution/interpreter/runtime.rb#L135-L144

It will take some work, but I think it will work just fine. 

I'm not exactly sure _how_ the runtime is supposed to support cases like this, though: 

```ruby
{
  field1 
  ... @defer {
    field1
  }
}
```

Should `field1` be present in the initial response, in a patch, or in both? I couldn't find anything in the spec to inform this. 

For now, I'm going to return it in the initial response and we'll see how it goes. If we have to rework it later, we can.

**Comment by rmosolgo** [comment]
I started taking a look over in https://github.com/rmosolgo/graphql-ruby/pull/3499. It will require a little bit of finesse to get all those pieces working together. In the meantime, I'd like to prioritize something that you can try in your own app. Are you using `GraphQL::Dataloader` and/or `GraphQL::Batch`?

**Comment by fazo96** [comment]
Thanks for diving into it so quickly! We only started adopting graphql last week and it's a big app. At the moment we are still fleshing out the graphql types and just focusing on getting it to work.

One of the reasons we decided to adopt graphql is that we had very big JSON payloads in teh REST API that in some cases took a very long time to generate on the backend. The idea was that by using GraphQL and a Relay client we would have a lot more flexibility in organizing how data is fetched and we were right.

`@defer` would just make it even easier to mark some sections of the response as non blocking to avoid the page being in a loading state while getting data that isn't immediately needed

Eventually I would have tried `@stream` and `@stream_connection` directives too, but we don't have as strong a need for those as well and they seem even less documented than `@defer`

We aren't using any batch loading system yet but I am sure we would benefit from it

**Comment by rmosolgo** [comment]
Glad to hear it's working well so far! That certainly sounds like the sweet spot for GraphQL and `@defer`. 

Things are _basically_ working over on #3499 and the performance is looking good. I'll double-check that `@defer` will work as expected and cut new gem versions soon with the required changes, then follow up here.

**Comment by rmosolgo** [comment]
👋  I just released 1.12.12 which _should_ fix this issue. If you end up using `GraphQL::Dataloader`, you'll need graphql-pro 1.18.0, too. Please let me know if you run into any more trouble with those versions!

**Comment by fazo96** [comment]
Hey sorry for the late response. I finally got around to trying `@defer` again and it works perfectly on fragment expansions. Thanks a lot for the quick turnaround!
